locore: don't mess with CPUECTLR_EL1 except on Cortex-A53,-A57,-A72

richlowe opened the pull request:
This should follow all the CPU manuals as to what we need to do (and not do) here, and do so tightly enough that qemu is happy with us now where it wasn't before.

@citrus-it @hadfl can you see how this goes on raspberry pi, and hvf?  It should be fine and/or invisibly better, I hope?  (it probably won't help showing your native CPU through qemu under hvf though, until we support the larger address space).

richlowe committed:
locore: don't mess with CPUECTLR_EL1 except on Cortex-A53,-A57,-A72

On real hardware, we probably "just" end up writing a now-RES0 bit of
CPUECTLR_EL1, or trapping because we never arranged for EL1 to be abble to
access the register.

In qemu however, either because of the access control, or maybe qemu doesn't
implement the register, we trap on the msr and immediately fail that PE.

While here:

  - Switch to symbolic names in locore.S
  - Tighten up the ASM in locore.S and srt0.S
  - Apply this code to Cortex-A72, which the manual at least implies
    is either necessary or a no-op in both locore.S and srt0.S, including
    access to CPUECTLR on -A72.
  - Remove the unused C asm inlines to access CPUECTLR_EL1
  - Use ENTRY/SET_SIZE for `secondary_vec_start`

richlowe commented:
With this, qemu boots as cortex-a76, and neoverse-n1 (but not -n2 yet)

citrus-it commented:
No visible difference under hvf, same boot speed as far as I can tell, four cores showing up as before, and I can see four host cores being active too.
```
The illumos Project     arm64/cortex-a76-0-gf7e90717f   Jun. 29, 2024
illumos development build:  2024-Jun-29 [illumos-gate]
root@braich:~# psrinfo -pv
The physical processor has 4 virtual processors (0-3)
  AArch64 (midr 610f0000 revidr 00000000)
        Apple Unknown [0] @ 1000 MHz
```

richlowe committed:
unix: handle chips with FEAT_CCIDX

When FEAT_CCIDX the locations of CCSIDR.NumSets and CCSIDR.Associativity
within the register change incompatibly.  Handle this.

richlowe commented:
For ease of sanity, since it's in the same theme.  I added the commit here that allows the virt platform's emulated neoverse-v1, neoverse-n2 and cortex-a710 work.

They have further problems (not least the fact they don't believe they have this feature, by the time they've booted), but this gets them into userland and running seemingly well, since the feature-mishap has no practical effect at the moment.

@citrus-it sorry, could you check this again?
